lib/oga/xml/lexer.rb

```diff
@@ -512,8 +512,11 @@ def on_element_end(name = nil)
           end
         end
 
-        add_token(:T_ELEM_END)
+        # Prevents a superfluous end tag of a self-closing HTML tag from
+        # closing its parent element prematurely
+        return if html? and name != nil and name != current_element
```

Comment:
For one liners I prefer the use of `&&` and `||`. It's not a big deal here because precedence doesn't doesn't matter, but I'd prefer to keep it consistent with any other one-liners there might be. For whatever reason I tend to write one-liners where the different precedence of `and` / `or` actually matters, hence I prefer to use `&&` and `||` for them.

Also, the `name != nil` can just be `name` leaving you with:

    return if html? && name && name != current_element

Other than that, great work!